sunpy/net/dataretriever/sources/gong_synoptic.py

```diff
@@ -0,0 +1,172 @@
+import re
+import os
+import gzip
+from pathlib import Path
+
+from urllib.error import HTTPError
+from urllib.request import urlopen
+
+from bs4 import BeautifulSoup
+
+from parfive import Downloader
+
+from ..client import GenericClient
+from sunpy.util.scraper import Scraper
+from sunpy.time import parse_time
+
+__all__ = ['GongSynopticClient']
+
+
+class GongSynopticClient(GenericClient):
+    """
+    Provides access to the merged data product for NSO-GONG synoptic maps
+    `archive <gong2.nso.edu/oQR/zqs/>`__
+
+    Examples
+    --------
+
+    >>> from sunpy.net import Fido, attrs as a
+    >>> results = Fido.search(a.Time("2019/12/31 22:00", "2020/1/1 02:00"),
+    ...                       a.Instrument('GONG'))  #doctest: +REMOTE_DATA
+    >>> results  #doctest: +REMOTE_DATA
+    <sunpy.net.fido_factory.UnifiedResponse object at ...>
+    Results from 1 Provider:
+    <BLANKLINE>
+    3 Results from the GongSynopticClient:
+         Start Time           End Time      Source Instrument Wavelength
+    ------------------- ------------------- ------ ---------- ----------
+    2019-12-31 22:00:00 2020-01-01 02:00:00    NSO       GONG        nan
+    2019-12-31 22:00:00 2020-01-01 02:00:00    NSO       GONG        nan
+    2019-12-31 22:00:00 2020-01-01 02:00:00    NSO       GONG        nan
+    <BLANKLINE>
+    <BLANKLINE>
+
+    """
+
+    def _get_url_for_timerange(self, timerange, **kwargs):
+        """
+        Return URL(s) for corresponding timerange.
+
+        Parameters
+        ----------
+        timerange : `~sunpy.time.TimeRange`
+            The time range you want the files for.
+
+        Returns
+        -------
+        `list`
+            The URL(s) for the corresponding timerange.
+        """
+        prefix = 'https://gong2.nso.edu/oQR/zqs/%Y%m/mrzqs%y%m%d/'
+        file_pattern = 'mrzqs\d{6}t\d{4}c\d+_\d+.fits.gz'
+        pref_scrap = Scraper(prefix)
+        directories = pref_scrap.range(timerange)
+        filesurls = []
+        for directory in directories:
+            try:
+                opn = urlopen(directory)
+                try:
+                    soup = BeautifulSoup(opn, "html.parser")
+                    for link in soup.find_all("a"):
+                        href = link.get("href")
+                        if re.match(file_pattern, href):
+                            # extracting timestamp from the filename
+                            timestring = href[5:16]
+                            tstamp = parse_time('20'+timestring.upper()+'00')
+                            if tstamp >= timerange.start and tstamp <= timerange.end:
+                                fullpath = directory + href
+                                filesurls.append(fullpath)
```

Comment:
Shouldn't all that be "done" by the scrapper? I believe you can put a wildcard if you don't know how it goes (like the last digits)... Ok, you can't! but with my new pr #4107

Reply:
This PR will be very helpful. I'll change the code to depend on #4107 after it is merged.